### PR TITLE
fd_shred: tighten parse constraints, other small changes

### PR DESCRIPTION
**Pull request by ptaffet-jump**
Basically rewrote the parsing function, added a fuzzer. 

Other assorted small changes, mostly changes that Richie made, plus a few fixes, many of which Josh also made

Note: Way back when, Tom suggested some changes on #123 that we never made, but it's nothing critical.

### CODE REVIEW

File: src/ballet/shred/fd_shred.c [hunk 1]
```diff
@@ -1,15 +1,125 @@
...
 
 fd_shred_t const *
-fd_shred_parse( uchar const * buf ) {
+fd_shred_parse( uchar const * const buf,
+                ulong         const sz ) {
+  /* Reinterpret size as long: `sz` cannot realistically underflow.
+     `sz1` is the size of the entire shred. */
+  long const sz1 = (long)sz;
+
+  /* Initial bounds check */
```
Comment (mmcgee-jump) [inline src/ballet/shred/fd_shred.c:10]: Maybe just move this above? Then don't have the underflow question

File: src/ballet/shred/fd_shred.c [hunk 2]
```diff
@@ -1,15 +1,125 @@
...
+  if( FD_UNLIKELY( sz1<0x58L ) ) return NULL;
+
   fd_shred_t const * shred = (fd_shred_t *)buf;
-  /* Validate shred type */
-  uchar shred_type = fd_shred_type( shred->variant );
-  if( FD_UNLIKELY( shred_type!=FD_SHRED_TYPE_MERKLE_DATA &&
-                   shred_type!=FD_SHRED_TYPE_MERKLE_CODE &&
+
+  /* Validate shred type.
+     Safe to access because `variant` ends at 0x41, which is <= 0x58 */
```
Comment (mmcgee-jump) [inline src/ballet/shred/fd_shred.c:16]: Any way to add a `static_assert( offsetof(shred, variant), 0x41)` here? Just a bit paranoid with parsing code.

File: src/ballet/shred/fd_shred.c [hunk 3]
```diff
@@ -1,15 +1,125 @@
...
+
+  /* Bounds check merkle data
+     `sz2` is the shred size excluding Merkle proof data. */
+  long const sz2 = sz1 - (long)fd_shred_merkle_sz( shred->variant );
+  if( FD_UNLIKELY( sz2<0L ) ) return NULL;
+
+  /* `sz3` is the payload size of the shred. */
+  if( FD_LIKELY( type & (FD_SHRED_TYPE_LEGACY_DATA|FD_SHRED_TYPE_MERKLE_DATA) ) ) {
+    /* Bounds check actual size
+       TODO Verify whether this includes the merkle (sz1) or not (sz2) */
```
Comment (mmcgee-jump) [inline src/ballet/shred/fd_shred.c:32]: Noting the todo :)

**Review by mmcgee-jump — DISMISSED** [review]

File: src/ballet/shred/fd_shred.c [hunk 4]
```diff
@@ -1,15 +1,125 @@
...
+    if( FD_UNLIKELY( shredder->shred_cnt == 0U ) )
+      return (shredder->result = -FD_SHRED_EPIPE);
+
+    fd_shred_t const * shred = *shredder->shreds;
+
+    /* Sanity check: Type must be data shred */
+    uchar shred_type = fd_shred_type( shred->variant );
+    if( FD_UNLIKELY( shred_type!=FD_SHRED_TYPE_LEGACY_DATA
+                  && shred_type!=FD_SHRED_TYPE_MERKLE_DATA ) )
+      return -FD_SHRED_EINVAL;
```
[the remark this answers is not in the document]
Comment (ptaffet-jump) [reply src/ballet/shred/fd_shred.c]: I removed the deshredder code because we don't need it right now.

File: src/ballet/shred/fd_shred.c [hunk 5]
```diff
@@ -1,15 +1,60 @@
...
+  if( FD_LIKELY( type & (FD_SHRED_TYPE_LEGACY_DATA|FD_SHRED_TYPE_MERKLE_DATA) ) ) {
+    if( FD_UNLIKELY( shred->data.size<header_sz ) ) return NULL;
+    payload_sz = (ulong)shred->data.size - header_sz; /* between 0 and USHORT_MAX */
+    if( FD_UNLIKELY( (variant!=FD_SHRED_TYPE_LEGACY_DATA) & (sz<FD_SHRED_MIN_SZ) ) ) return NULL;
+
+    /* legacy data shreds might be shorter than the normal
+       FD_SHRED_MIN_SZ, but they don't have Merkle proofs, so everything
+       after the payload is zero-padding/ignored.  On the other hand,
+       Merkle data shreds might have some zero-padding, but anything
+       between [FD_SHRED_MIN_SZ, sz) is junk.  The Merkle proof is in
```
Comment (ptaffet-jump) [inline src/ballet/shred/fd_shred.c:40]:
```suggestion
       between [FD_SHRED_MIN_SZ, sz) is junk.  The Merkle proof is not in
```

**Review by mmcgee-jump — APPROVED** [review]

**Pull request merged**